### PR TITLE
Plugins: Add support for platform specific plugins

### PR DESCRIPTION
**Pull request by rjernst**
This commit adds support for plugins having a platform specific variant.
It also adds unit tests for all official and maven urls.

### CODE REVIEW

File: distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java [hunk 1]
```diff
@@ -253,6 +242,60 @@ private Path download(Terminal terminal, String pluginId, Path tmpDir) throws Ex
...
+                "https://staging.elastic.co/%1$s-%2$s/downloads/elasticsearch-plugins/%3$s", version, stagingHash, pluginId);
+        } else {
+            baseUrl = String.format(Locale.ROOT,
+                "https://artifacts.elastic.co/downloads/elasticsearch-plugins/%1$s", pluginId);
+        }
+        final String platformUrl = String.format(Locale.ROOT, "%1$s/%2$s-%3$s-%4$s.zip", baseUrl, pluginId, platform, version);
+        if (urlExists(terminal, platformUrl)) {
+            return platformUrl;
+        }
+        return String.format(Locale.ROOT, "%1$s/%2$s-%3$s.zip", baseUrl, pluginId, version);
```
Comment (jasontedor) [inline distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java]: It looks like we do not need the argument index modifiers in any of the format strings in this method anymore?

File: distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java [hunk 2]
```diff
@@ -253,6 +242,60 @@ private Path download(Terminal terminal, String pluginId, Path tmpDir) throws Ex
...
+    /**
+     * Returns {@code true} if the given url exists, and {@code false} otherwise.
+     *
+     * The given url must be {@code https} and existing means a {@code HEAD} request returns 200.
+     */
+    // pkg private for tests to manipulate
+    @SuppressForbidden(reason = "Make HEAD request using URLConnection.connect()")
+    boolean urlExists(Terminal terminal, String urlString) throws IOException {
+        terminal.println(VERBOSE, "Checking if url exists: " + urlString);
+        assert urlString.startsWith("https://") : "Only http urls can be checked";
```
Comment (jasontedor) [inline distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java]: The assert message should be `https` not `http`!

Comment (jasontedor) [inline distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java]: And maybe it's better to construct the URL and just assert that `"https".equals(url.getProtocol())`?

File: distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java [hunk 3]
```diff
@@ -253,6 +242,60 @@ private Path download(Terminal terminal, String pluginId, Path tmpDir) throws Ex
...
+        URL url = new URL(urlString);
+        HttpURLConnection urlConnection = (HttpURLConnection)url.openConnection();
```
Comment (jasontedor) [inline distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java]: Nit: space between the cast operator and the target.

File: distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java [hunk 4]
```diff
@@ -253,6 +242,60 @@ private Path download(Terminal terminal, String pluginId, Path tmpDir) throws Ex
...
+        urlConnection.addRequestProperty("User-Agent", "elasticsearch-plugin-installer");
+        urlConnection.setRequestMethod("HEAD");
+        urlConnection.connect();
+        return urlConnection.getResponseCode() == 200;
```
Comment (jasontedor) [inline distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java]: What if there's a redirect?

File: distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java [hunk 5]
```diff
@@ -253,6 +242,60 @@ private Path download(Terminal terminal, String pluginId, Path tmpDir) throws Ex
...
+    private String getMavenUrl(Terminal terminal, String[] coordinates, String platform) throws IOException {
+        final String groupId = coordinates[0].replace(".", "/");
+        final String artifactId = coordinates[1];
+        final String version = coordinates[2];
+        final String baseUrl = String.format(Locale.ROOT, "https://repo1.maven.org/maven2/%1$s/%2$s/%3$s", groupId, artifactId, version);
+        final String platformUrl = String.format(Locale.ROOT, "%1$s/%2$s-%3$s-%4$s.zip", baseUrl, artifactId, platform, version);
+        if (urlExists(terminal, platformUrl)) {
+            return platformUrl;
+        }
+        return String.format(Locale.ROOT, "%1$s/%2$s-%3$s.zip", baseUrl, artifactId, version);
```
Comment (jasontedor) [inline distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java]: Same thing here, I don't think the argument index modifiers are needed?

**Review by jasontedor — COMMENTED** [review]
I left some comments, the main one is about the redirects.

File: distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java (hunk 4 above)
Comment (rjernst) [inline distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java]: We don't support redirects right now. The code in downloadZip does a direct connection, no redirect handling.

Comment (jasontedor) [inline distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java]: Ah, okay. Then I'm good with it as is; we can consider redirects separately (if ever).

**Review by jasontedor — APPROVED** [review]
LGTM.

**Pull request merged**